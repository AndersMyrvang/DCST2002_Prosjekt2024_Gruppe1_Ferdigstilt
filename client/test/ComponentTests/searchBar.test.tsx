import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter, Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import SearchBar from '../../src/components/searchBar';

global.fetch = jest.fn();

const mockFetch = fetch as jest.Mock;

describe('SearchBar Component', () => {
  const onSearchMock = jest.fn();

  const setup = () => {
    const history = createMemoryHistory();
    return {
      history,
      ...render(
        <Router history={history}>
          <SearchBar onSearch={onSearchMock} />
        </Router>,
      ),
    };
  };

  beforeEach(() => {
    mockFetch.mockClear();
    onSearchMock.mockClear();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders the search bar input and search icon', () => {
    const { asFragment } = setup();

    expect(screen.getByPlaceholderText('Search...')).toBeInTheDocument();
    expect(screen.getByText('🔍')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  test('does not fetch suggestions for queries shorter than 3 characters', async () => {
    setup();

    const input = screen.getByPlaceholderText('Search...');
    fireEvent.change(input, { target: { value: 'te' } });

    await waitFor(() => expect(mockFetch).not.toHaveBeenCalled());

    expect(screen.queryByText('Player:')).not.toBeInTheDocument();
  });

  test('fetches and displays suggestions when typing in the search bar', async () => {
    mockFetch.mockResolvedValueOnce({
      json: async () => ({
        suggestions: [
          { type: 'Player', id: 1, name: 'Player One' },
          { type: 'Team', id: 2, name: 'Team Two' },
        ],
      }),
    });

    const { asFragment } = setup();

    const input = screen.getByPlaceholderText('Search...');
    fireEvent.change(input, { target: { value: 'pla' } });

    await waitFor(() => expect(mockFetch).toHaveBeenCalledWith('/api/search/suggestions?q=pla'));

    expect(await screen.findByText('Player: Player One')).toBeInTheDocument();
    expect(await screen.findByText('Team: Team Two')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  test('navigates to the correct path when a suggestion is clicked', async () => {
    const { history } = setup();

    mockFetch.mockResolvedValueOnce({
      json: async () => ({
        suggestions: [
          { type: 'Player', id: 1, name: 'Player One' },
          { type: 'Team', id: 2, name: 'Team Two' },
        ],
      }),
    });

    const input = screen.getByPlaceholderText('Search...');
    fireEvent.change(input, { target: { value: 'pla' } });

    await waitFor(() => screen.findByText('Player: Player One'));

    const suggestion = screen.getByText('Player: Player One');
    fireEvent.click(suggestion);

    expect(history.location.pathname).toBe('/player/1');
  });

  test('clears suggestions when search is triggered', async () => {
    mockFetch.mockResolvedValueOnce({
      json: async () => ({
        suggestions: [{ type: 'Player', id: 1, name: 'Player One' }],
      }),
    });

    const { asFragment } = setup();

    const input = screen.getByPlaceholderText('Search...');
    fireEvent.change(input, { target: { value: 'pla' } });

    await waitFor(() => screen.findByText('Player: Player One'));

    const searchIcon = screen.getByText('🔍');
    fireEvent.click(searchIcon);

    await waitFor(() => expect(screen.queryByText('Player: Player One')).not.toBeInTheDocument());

    expect(asFragment()).toMatchSnapshot();
  });

  test('handles API errors gracefully', async () => {
    mockFetch.mockRejectedValueOnce(new Error('API error'));

    setup();

    const input = screen.getByPlaceholderText('Search...');
    fireEvent.change(input, { target: { value: 'pla' } });

    await waitFor(() => expect(mockFetch).toHaveBeenCalledWith('/api/search/suggestions?q=pla'));

    expect(screen.queryByText('Player: Player One')).not.toBeInTheDocument();
    expect(screen.queryByText('Team: Team Two')).not.toBeInTheDocument();
  });
});
