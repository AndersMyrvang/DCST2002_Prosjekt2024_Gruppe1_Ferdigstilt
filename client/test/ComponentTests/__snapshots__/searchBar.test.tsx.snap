// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component clears suggestions when search is triggered 1`] = `
<DocumentFragment>
  <div
    class="search-bar"
  >
    <input
      placeholder="Search..."
      type="text"
      value="pla"
    />
    <span
      class="search-icon"
    >
      🔍
    </span>
  </div>
</DocumentFragment>
`;

exports[`SearchBar Component fetches and displays suggestions when typing in the search bar 1`] = `
<DocumentFragment>
  <div
    class="search-bar"
  >
    <input
      placeholder="Search..."
      type="text"
      value="pla"
    />
    <span
      class="search-icon"
    >
      🔍
    </span>
    <div
      class="suggestions-dropdown"
    >
      <div
        class="suggestion-item"
      >
        Player: Player One
      </div>
      <div
        class="suggestion-item"
      >
        Team: Team Two
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SearchBar Component renders the search bar input and search icon 1`] = `
<DocumentFragment>
  <div
    class="search-bar"
  >
    <input
      placeholder="Search..."
      type="text"
      value=""
    />
    <span
      class="search-icon"
    >
      🔍
    </span>
  </div>
</DocumentFragment>
`;
